Passing a byte value as defaultable to a Ballerina function gives Bad, sad error.

**Description:**
$Subject

**Steps to reproduce:**
```
public function main() {
     testFunc(10);
}

function testFunc(byte b, byte c = 10) {
     byte a = <byte> (b + c);
}
```

Error log:
```
[2021-03-12 14:30:44,437] SEVERE {b7a.log.crash} - null 
java.lang.ArrayIndexOutOfBoundsException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
	at java.base/java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:678)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:737)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:661)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generateModuleClasses(JvmPackageGen.java:452)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.generate(JvmPackageGen.java:438)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:89)
	at org.wso2.ballerinalang.compiler.bir.codegen.CodeGenerator.generate(CodeGenerator.java:67)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:267)
	at io.ballerina.projects.ModuleContext.generateCodeInternal(ModuleContext.java:398)
	at io.ballerina.projects.ModuleCompilationState$4.generatePlatformSpecificCode(ModuleCompilationState.java:132)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:327)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:145)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:131)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:106)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:129)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:105)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.BuildCommand.execute(BuildCommand.java:251)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 0 out of bounds for length 0
	at org.objectweb.asm.Frame.merge(Frame.java:1268)
	at org.objectweb.asm.Frame.merge(Frame.java:1233)
	at org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1607)
	at org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1543)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.genJMethodForBFunc(MethodGen.java:259)
	at org.wso2.ballerinalang.compiler.bir.codegen.methodgen.MethodGen.generateMethod(MethodGen.java:154)
	at org.wso2.ballerinalang.compiler.bir.codegen.JvmPackageGen.lambda$generateModuleClasses$0(JvmPackageGen.java:502)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1746)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```
**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**
<!-- Any related issues such as sub tasks, issues reported in other repositories (e.g component repositories), similar problems, etc. -->

**Suggested Labels (optional):**
Type/Bug, Team/jBallerina

**Suggested Assignees (optional):**
<!--Optional comma separated list of suggested team members who should attend the issue. Non committers can’t assign issues to assignees, so this will help issue creators who are not a committer to suggest possible assignees-->

--- a/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java
+++ b/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmTerminatorGen.java
@@ -203,7 +203,7 @@ public class JvmTerminatorGen {
     }
 
     private void loadDefaultValue(MethodVisitor mv, BType bType) {
-        if (TypeTags.isIntegerTypeTag(bType.tag) || bType.tag == TypeTags.BYTE) {
+        if (TypeTags.isIntegerTypeTag(bType.tag)) {
             mv.visitInsn(LCONST_0);
             return;
         } else if (TypeTags.isStringTypeTag(bType.tag) || TypeTags.isXMLTypeTag(bType.tag)) {
@@ -215,6 +215,7 @@ public class JvmTerminatorGen {
             case TypeTags.FLOAT:
                 mv.visitInsn(DCONST_0);
                 break;
+            case TypeTags.BYTE:
             case TypeTags.BOOLEAN:
                 mv.visitInsn(ICONST_0);
                 break;

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/FunctionsWithDefaultableArguments.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/functions/FunctionsWithDefaultableArguments.java
@@ -32,6 +32,7 @@ import org.ballerinalang.test.CompileResult;
 import org.testng.Assert;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
 /**
@@ -186,6 +187,19 @@ public class FunctionsWithDefaultableArguments {
         bValueArray = (BValueArray) returns[1];
         Assert.assertEquals(bValueArray.getRefValue(0).stringValue(), "given");
         Assert.assertEquals(((BInteger) bValueArray.getRefValue(1)).intValue(), 200);
+    }
+
+    @Test(dataProvider = "functionsWithDefaultByteValues")
+    public void testFuncWithDefaultByteParamValues(String function) {
+        BRunUtil.invoke(result, function);
+    }
+
+    @DataProvider
+    public  Object[] functionsWithDefaultByteValues() {
+        return new String[] {
+                "testFuncWithDefaultByteValue",
+                "testDefaultByteValueInFunctionPointers"
+        };
     }
 
     @Test

--- a/tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal
@@ -14,6 +14,7 @@
 // specific language governing permissions and limitations
 // under the License.
 import ballerina/jballerina.java;
+import ballerina/test;
 
 int GLB = 0;
 
@@ -188,6 +189,22 @@ function testDefaultObject() returns [[string, int], [string, int]] {
 
 function foo7(string s, float b = 1.1, FooObject o = new("default", 100)) returns [string, float, FooObject] {
     return [s, b, o];
+}
+
+function testFuncWithDefaultByteValue() {
+    byte b = funcWithDefaultByteValue();
+    test:assertEquals(b, 10);
+}
+
+function testDefaultByteValueInFunctionPointers() {
+    function (byte) returns byte fp = funcWithDefaultByteValue;
+    byte b = fp(15);
+
+    test:assertEquals(b, 15);
+}
+
+function funcWithDefaultByteValue(byte b = 10) returns byte {
+    return b;
 }
 
 function testFuncWithAsyncDefaultParamExpression() returns string {

--- a/tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal
@@ -198,9 +198,11 @@ function testFuncWithDefaultByteValue() {
 
 function testDefaultByteValueInFunctionPointers() {
     function (byte b = 5) returns byte fp = funcWithDefaultByteValue;
-
     test:assertEquals(fp(15), 15);
     test:assertEquals(fp(funcWithDefaultByteValue()), 10);
+
+    //TODO : Enable this after merge #31589
+    //test:assertEquals(fp(), 5);
 }
 
 function funcWithDefaultByteValue(byte b = 10) returns byte {

--- a/tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal
+++ b/tests/jballerina-unit-test/src/test/resources/test-src/functions/functions_with_default_parameters.bal
@@ -192,15 +192,15 @@ function foo7(string s, float b = 1.1, FooObject o = new("default", 100)) return
 }
 
 function testFuncWithDefaultByteValue() {
-    byte b = funcWithDefaultByteValue();
-    test:assertEquals(b, 10);
+    test:assertEquals(funcWithDefaultByteValue(), 10);
+    test:assertEquals(funcWithDefaultByteValue(15), 15);
 }
 
 function testDefaultByteValueInFunctionPointers() {
-    function (byte) returns byte fp = funcWithDefaultByteValue;
-    byte b = fp(15);
+    function (byte b = 5) returns byte fp = funcWithDefaultByteValue;
 
-    test:assertEquals(b, 15);
+    test:assertEquals(fp(15), 15);
+    test:assertEquals(fp(funcWithDefaultByteValue()), 10);
 }
 
 function funcWithDefaultByteValue(byte b = 10) returns byte {

